refactor: ♻️ update type hints for all documentation

onuralpszr opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->


## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
Modernizes type hints and docs across Ultralytics to use Python’s built‑in generics and union syntax, improving clarity and consistency without changing behavior. ✅

### 📊 Key Changes
- Standardized type annotations in docs, examples, and docstrings:
  - `List` → `list`, `Dict` → `dict`, `Tuple` → `tuple`, `Optional[T]` → `T | None`
  - Consistent lowercase wording in descriptions (e.g., “list of …”)
- Updated example code signatures and comments across multiple examples (ONNX Runtime, TFLite, OpenCV, RTDETR, action recognition, etc.)
- Minor doc fixes and clarity improvements in:
  - Ultralytics HUB, datasets, solutions, callbacks, predict/val modes, SAM/SAM2 modules, and configuration helpers
  - Vertex AI Docker guide example: `run_inference` and Pydantic models updated to `dict[...]` and `| None`
- No functional changes to core logic; primarily documentation and typing consistency

### 🎯 Purpose & Impact
- Improves readability and consistency for users and contributors 🧹
- Aligns docs with modern Python typing (PEP 585/604) for better editor/IDE support 💡
- Reduces confusion in examples and API docs by using a single, consistent style 📘
- No breaking changes to APIs or model behavior; training, inference, and exports remain unchanged 🔒
- Note: If copying example code, Python 3.9+ is recommended to support the newer type syntax ⚠️

onuralpszr committed:
refactor: ♻️ update  type hints for all docs

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant commented:
👋 Hello @onuralpszr, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help speed things up — an Ultralytics engineer will also review and assist you shortly 😊

To ensure a smooth review and merge process, please go through the checklist below:

- ✅ Define a Purpose: Clearly explain the purpose and scope of your refactor in the PR description (modernizing type hints and docs with built-in generics and union syntax), and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear and follow project conventions.
- ✅ Synchronize with Source: Confirm your PR is up to date with the `ultralytics/ultralytics` `main` branch. If it’s behind, click “Update branch” or run `git pull` and `git merge main` locally.
- ✅ Ensure CI Checks Pass: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address them.
- ✅ Update Documentation: Since this PR updates type hints and docs across examples and guides, double-check that the updated syntax renders correctly in our [documentation](https://docs.ultralytics.com/). Please ensure code blocks remain valid and consistent with Python 3.9+ recommendations noted in your summary.
- ✅ Add Tests: If any runtime-facing code paths or examples were adjusted beyond docs (e.g., example scripts), consider adding or updating tests where applicable, and confirm they pass locally.
- ✅ Sign the CLA: If this is your first Ultralytics contribution, please sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by commenting “I have read the CLA Document and I sign the CLA.”
- ✅ Minimize Changes: Keep changes to the minimum necessary for the documentation and typing consistency refactor. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional suggestions for this PR:
- 🔍 Consistency pass: Ensure consistent lowercase type names in all tables, docstrings, and comments (e.g., `list`, `dict`, `tuple`, `T | None`) and verify uniform usage across examples (ONNX Runtime, TFLite, OpenCV, RTDETR, SAM/SAM2, solutions, callbacks).
- 🧪 Example validation: For modified example scripts, run a quick local check to ensure imports, annotations, and docstrings remain aligned after type updates.
- 🧭 Branding check: Confirm references align with Ultralytics branding (e.g., YOLO11 naming where applicable and “Ultralytics HUB” formatting).
- 🧱 Docs build: If possible, build the docs locally to catch any formatting or indentation issues introduced in code fences or tables during type changes.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

glenn-jocher committed:
Update api.md

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher commented on docs/en/help/contributing.md:
````diff
@@ -135,7 +135,7 @@ When adding new functions or classes, include [Google-style docstrings](https://
         ```
         Returns:
-           (tuple): Tuple containing:
+           (tuple): tuple containing:
````
Ah, looks like we have a lot of accidental conversions here. First word of the sentence should stay capitalized. Maybe search and replace "): tuple" etc.

onuralpszr committed:
refactor: ♻️ Improve code readability and consistency and update typing.

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
refactor: ♻️ Update isinstance checks to use tuple syntax for type checking

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
fix: 🐞 mkdocs strict mode typing check fix

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

glenn-jocher committed:
Update minimum-reproducible-example.md

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Update solutions-args.md

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Update sam-auto-annotate.md

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher commented on docs/en/models/sam-2.md:
```diff
@@ -298,7 +298,7 @@ It offers three significant enhancements:
 | `max_obj_num`   | `3`           | `str`       | The preset maximum number of categories     |
 | `update_memory` | `False`       | `bool`      | Whether to update memory with new prompts   |
-| `obj_ids`       | `None`        | `List[int]` | List of object IDs corresponding to prompts |
+| `obj_ids`       | `None`        | `list[int]` | list of object IDs corresponding to prompts |
```
@onuralpszr we need to keep the start of any sentence capitalized, so the second list here should be List

onuralpszr replied:
Of course I was finishing code sections first around Optionals then I am going to handle docs parts as well. I am well aware

glenn-jocher commented on ultralytics/models/sam/modules/sam.py:
```diff
@@ -58,8 +58,8 @@ def __init__(
         mask_decoder: MaskDecoder,
-        pixel_mean: list[float] = (123.675, 116.28, 103.53),
-        pixel_std: list[float] = (58.395, 57.12, 57.375),
+        pixel_mean: list[float] = [123.675, 116.28, 103.53],
+        pixel_std: list[float] = [58.395, 57.12, 57.375],
```
@onuralpszr problem here is Python now wants immutable arguments, so I changed these form list to tuple a while back. A little annoying but means that we're not supposed to had default args as lists.

onuralpszr replied:
I will revert okay 👍

glenn-jocher committed:
Update object_counter.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Update streamlit_inference.py

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Use | 

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher commented on ultralytics/utils/export.py:
```diff
@@ -27,9 +27,9 @@ def export_onnx(
-        dynamic (bool | Dict, optional): Whether to enable dynamic axes.
+        input_names (list[str]): List of input tensor names.
+        output_names (list[str]): List of output tensor names.
+        dynamic (bool | dict, optional): Whether to enable dynamic axes.
```
@onuralpszr we need to check if Google-style docstrings use ", optional" or "| None" and then be consistent.

onuralpszr committed:
refactor: ♻️ type hints and documentation across multiple modules and remove optional and None

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge b50d3a45e58feb19372e80e6af79b786a89d6ebf into 9bd89be040f98688bde8a154622bccdc1c355d32

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
fix: 🐞 add future annotations import for type hinting support

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
fix: 🐞 change pixel_mean and pixel_std default values from list to tuple for consistency

onuralpszr committed:
fix: 🐞 update documentation capital letter typo fix

glenn-jocher commented:
@onuralpszr just merged https://github.com/ultralytics/actions/pull/508 that will actually update all type hints to 3.9+ automatically.

glenn-jocher committed:
Add newline to trigger 3.9 hints actions

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

onuralpszr commented:
> @onuralpszr just merged [ultralytics/actions#508](https://github.com/ultralytics/actions/pull/508) that will actually update all type hints to 3.9+ automatically.

Thank you, I already used previously like this (FA) for future additions because we are using py38


```bash
ruff format \
--line-length 120 \
. || true
ruff check \
--fix \
--unsafe-fixes \
--extend-select I,D,UP,F,FA \
--target-version py39 \
--ignore D100,D104,D203,D205,D212,D213,D401,D406,D407,D413 \
```

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
@onuralpszr hmm, seems like F and FA are missing from actions then?

glenn-jocher committed:
fix indent

Signed-off-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher commented:
What a crazy PR. BTW @onuralpszr I just realized since I enabled 3.9 in https://github.com/ultralytics/actions/pull/508 then hint changes might start happening in all PRs with new commits, so we should try to get this merged to main as soon as possible. Then we can update all PR branches to get ahead of conflicts.

onuralpszr commented:
> What a crazy PR. BTW @onuralpszr I just realized since I enabled 3.9 in [ultralytics/actions#508](https://github.com/ultralytics/actions/pull/508) then hint changes might start happening in all PRs with new commits, so we should try to get this merged to main as soon as possible. Then we can update all PR branches to get ahead of conflicts.

Let's revert it, after this merge let's re-enable it. That would be better for all of us.

glenn-jocher commented:
@onuralpszr but I'm not sure what will happen with Python 3.8 targets with these 3.9 hints. I'm kind of confused that actions isn't reverting all of these to 3.8. What do you think is happening?

onuralpszr commented:
> Ok I'll revert actions back to 3.8 targets then untill we get this merged into main and the PR branches.

I already opened a PR please check it out :) 

https://github.com/ultralytics/actions/pull/511

onuralpszr commented:
> @onuralpszr thank you, merged!

I also send you a small request in slack, could you please check if it is okay.

glenn-jocher commented:
@ultralytics/run-ci

onuralpszr commented on ultralytics/trackers/track.py:
```diff
@@ -75,7 +75,7 @@ def on_predict_postprocess_end(predictor: object, persist: bool = False) -> None
     Args:
         predictor (object): The predictor object containing the predictions.
-        persist (bool, optional): Whether to persist the trackers if they already exist.
+        persist (bool | None): Whether to persist the trackers if they already exist.
```
I revert my commits so "optional" is here

onuralpszr committed:
fix: 🐞 update type hints to use string annotations for better compatibility for python3.8

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge 39c0aeb7d0359881c23c66c16228a60ccf752ebe into 9bd89be040f98688bde8a154622bccdc1c355d32

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

onuralpszr committed:
refactor: ♻️ remove TYPE_CHECKING imports and simplify type hints across multiple files and keep backward compatible for py38 and formatter

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge f0fa7bdb5d1436220767d593331ad688694e51d1 into 9bd89be040f98688bde8a154622bccdc1c355d32

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
fix: 🐞 small typo fixes

onuralpszr committed:
refactor: ♻️ update small type hints

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr commented:
@ultralytics/run-ci

onuralpszr commented:
@glenn-jocher I think that is it.  We made a lot of changes and I reverted last auto optional replacement I made and I also revert TYPE_CHECKING because I realized python3.8 cannot handle and I tried to use "single quote"  handle properly but formatter is removing it, so I reverted that one as as well. 

A lot of them fixing "List" and "Tuple"  cases in documentation and docstrings . Also fixing some typing for list in code to make sure they all align. 

I made slow run and PR runs all passed. It should be good to go. 

Thank you for your help in documentation changes.

onuralpszr committed:
Merge branch 'main' into typing/updates-v2

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
refactor: ♻️ update type hints for confidence and IoU thresholds for example code

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into typing/updates-v2
